Account Recovery: Add the lost password page.

omarjackman opened the pull request:
# Account Recovery: add a lost password page

This pull request adds a new logged out page to calypso that is aimed at replacing https://wordpress.com/wp-login.php?action=lostpassword. This page doesn't actually do anything since a redux state tree hasn't been implemented yet to support any actions.
### How to test
#### Functionality
1. Navigate to http://calypso.localhost:3000/account-recovery
2. Enter a username
3. Click "Get new password"
4. Notice that the form is disabled
#### Responsiveness
1. Navigate to http://calypso.localhost:3000/account-recovery
2. Resize browser so that its small
3. Notice that the form is responsive
#### automated tests

Run `npm run test-client client/account-recovery/lost-password`
### What to expect

![screen shot 2016-10-25 at 7 01 46 pm](https://cloud.githubusercontent.com/assets/1854440/19707538/23ecc9b2-9ae6-11e6-8fd8-744413ee47ee.png)

![screen shot 2016-10-25 at 7 05 22 pm](https://cloud.githubusercontent.com/assets/1854440/19707566/45c6374e-9ae6-11e6-8d47-eff63ff081c5.png)

![screen shot 2016-10-25 at 9 16 13 pm](https://cloud.githubusercontent.com/assets/1854440/19709904/54b01b46-9af8-11e6-87e9-a81b645ff2e2.png)

omarjackman commented:
cc @dllh

evilluendas commented:
Nice! Looks and runs as expected.

I've noticed a tiny typo in `We have a full guid to resetting your password.` though: there's a missing `e` in guide.

dllh commented on client/account-recovery/lost-password/index.jsx:
```diff
@@ -0,0 +1,106 @@
+					{ translate( 'Lost your password' ) }
+				</h2>
+
+				<p>{ translate( 'Follow theses simple steps to reset your account:' ) }</p>
```
Typo: theses

dllh commented:
Gave the code a quick look and nothing jumped out at me beyond a typo, but I'd prefer a quick look from @jordwest.

omarjackman commented:
@gwwar If your not to busy, do you mind giving this some review :)

gwwar commented on client/account-recovery/index.js:
```diff
@@ -0,0 +1,14 @@
+ */
+import controller from './controller';
+
+export default function() {
```
Might as well name this :D, it'll help if we export any other functions in this file

aduth replied:
Even if this is only client-side, I might suggest considering the isomorphic render router as a point of future compatibility if SSR is ever desired:

https://github.com/Automattic/wp-calypso/blob/master/docs/isomorphic-routing.md

As an example of a browser-only section using it:

https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/types/index.js

gwwar commented on client/account-recovery/controller.js:
```diff
@@ -0,0 +1,32 @@
+const ANALYTICS_PAGE_TITLE = 'Account Recovery';
+
+export default {
+	lostPassword( context ) {
```
Why are we exporting an object?

gwwar replied:
Guess this is a controller pattern?

gwwar commented on client/account-recovery/index.js:
```diff
@@ -0,0 +1,14 @@
+/**
+ * Internal dependencies
+ */
+import controller from './controller';
```
Alternatively we could just export the function, and import with:
```
export function lostPassword() {
//...
```
```
import { lostPassword } from './controller';
```

gwwar commented on client/account-recovery/controller.js:
```diff
@@ -0,0 +1,32 @@
+	lostPassword( context ) {
+		const basePath = context.path;
+
+		analytics.pageView.record( basePath, `${ ANALYTICS_PAGE_TITLE } > Lost Password` );
```
Note that we can use:
```
import PageViewTracker from 'lib/analytics/page-view-tracker';
<PageViewTracker path={ basePath } title={ `${ ANALYTICS_PAGE_TITLE } > Lost Password` } />
```

We generally also prefer using the redux analytics events over using the lib directly. (Makes component testing much easier

omarjackman replied:
Done in https://github.com/Automattic/wp-calypso/pull/8955/commits/1e728a3d37f46c00de24a6ba792e4378c06cb5c0

gwwar commented on client/account-recovery/controller.js:
```diff
@@ -0,0 +1,32 @@
+		analytics.pageView.record( basePath, `${ ANALYTICS_PAGE_TITLE } > Lost Password` );
+
+		renderWithReduxStore(
+			<div>
```
Let's at least move this into it's own function:

eg:
```
import { localize } from 'i18n-calypso';
const lostPasswordPage = localize( ( { translate } )  => {
    return (
        <div>
           <DocumentHead title={ translate( 'Lost Password ‹ Account Recovery' ) } />
            <PageViewTracker path={ basePath } title={ `${ ANALYTICS_PAGE_TITLE } > Lost Password` } />
           <LostPassword className="account-recovery__container" />
        </div>
    );
} );
```

aduth replied:
The [`<Main />` component](https://github.com/Automattic/wp-calypso/tree/master/client/components/main) would be more appropriate to use here in place of `div` for top-level controller element.

gwwar commented on client/wordpress-com.js:
```diff
@@ -193,6 +193,16 @@ sections.push( {
+		paths: [ '/account-recovery' ],
+		module: 'account-recovery',
+		secondary: false,
+		enableLoggedOut: true,
```
I think we'll need to add the SSR ready headers then to the top of the related files:
```
/** @ssr-ready **/
```
See: https://github.com/Automattic/wp-calypso/blob/master/docs/server-side-rendering.md anyone from hyperion should also know more

gwwar replied:
Or I could be mistaken if you don't need SSR, though I think most of our logged out pages in Calypso are.

omarjackman replied:
@gwwar / @aduth Is it enough to just not include `/** @ssr-ready **/` so that this doesn't get rendered server side? I think since this does a redirect if the user is logged in and probably will do other stuff based on api calls then its probably not a good candidate.

aduth replied:
I don't think you'll need the pragma unless you've marked the section as `isomorphic: true` (cc @ockham to confirm)

gwwar replied:
Any SSR'd pages can't include files that don't have the ssr-ready headers. To enable SSR for a particular route, I believe there's more wiring involved.

gwwar commented on client/account-recovery/lost-password/index.jsx:
```diff
@@ -0,0 +1,106 @@
+	}
+
+	onSubmit = () => {
+		const userLogin = findDOMNode( this.refs.user_login ).value;
```
We should be able to avoid doing this.

gwwar replied:
Note, that we shouldn't use the mixin, but instead setup the handlers manually: https://facebook.github.io/react/docs/two-way-binding-helpers.html#valuelink-without-linkedstatemixin

Also see: https://facebook.github.io/react/docs/forms.html#examples

@aduth or @dmsnell do you know of any newer patterns we can use here?

aduth replied:
Based on the component's usage, I'd probably recommend making it a controlled input, setting value into state `onChange` and submitting/rendering using the state value.

Else you could have a parent component manage the value in its state (using on `onChange` prop here), or manage via Redux state (probably overkill given the amount of boilerplate involved).

omarjackman replied:
Done in https://github.com/Automattic/wp-calypso/pull/8955/commits/a712370dabb06914d74da9e7f2bd317b10b6c3db

aduth commented on client/account-recovery/lost-password/index.jsx:
```diff
@@ -0,0 +1,106 @@
+					<FormInput
+						className="lost-password__user-login-input"
+						ref="user_login"
+						id="user_login"
```
Do you need an `id` or `name` on this input? Best to avoid if possible, since `id` should be unique per page.

aduth commented on client/account-recovery/lost-password/index.jsx:
```diff
@@ -0,0 +1,106 @@
+				</p>
+
+				<Card>
+					<FormLabel htmlFor="user_login">{ translate( 'Username or Email' ) }</FormLabel>
```
Ah, I suppose this is what the ID is for. In which case, you can avoid `id` by having the `<FormLabel>` wrap the input.

omarjackman replied:
Done in https://github.com/Automattic/wp-calypso/pull/8955/commits/63d76fbfd02869f4ccfffd33df4aac5c061c2d22

aduth commented on client/account-recovery/lost-password/test/index.jsx:
```diff
@@ -0,0 +1,55 @@
+
+describe( 'LostPassword', () => {
+	it( 'should render as expected', () => {
+		const wrapper = shallow( <LostPassword className="test__test" translate={ noop } /> );
```
[Lodash's `_.identity`](https://lodash.com/docs/4.16.4#identity) might be a more useful stand-in for `translate` than `noop`.

omarjackman replied:
Done in https://github.com/Automattic/wp-calypso/pull/8955/commits/d8a935ca34fc623e1a449957df7cb858a65ea79a

aduth commented on client/account-recovery/lost-password/test/index.jsx:
```diff
@@ -0,0 +1,55 @@
+		expect( wrapper ).to.have.className( 'lost-password__container' );
+		expect( wrapper ).to.have.className( 'test__test' );
+		expect( wrapper ).to.have.state( 'isSubmitting' ).to.be.false;
+		expect( wrapper.find( '.lost-password__user-login-input' ).prop( 'disabled' ) ).to.not.be.ok;
```
`enzyme-chai` convenience for this could be:

```js
expect( wrapper.find( '.lost-password__user-login-input' ) ).to.not.have.prop( 'disabled' );
```

omarjackman replied:
Using this convenience means that I have to do complicated logic to add the `disabled` property
My code will look something like this:
```js
const disabledProp = isSubmitting ? { disabled: true } : {};
return <input { ...disabledProp }/>
```
instead of just `<input disabled={ isSubmitting } />`

aduth commented on client/account-recovery/lost-password/test/index.jsx:
```diff
@@ -0,0 +1,55 @@
+	} );
+
+	context( 'events', () => {
+		reactTapEventPlugin();
```
I think you won't need either of these for your tests, since you're not using `onTouchTap` in your component. Instead, you should be able to simulate clicks on a shallow rendered component without issue.

omarjackman replied:
Done in https://github.com/Automattic/wp-calypso/pull/8955/commits/211057a51aed592445966c930a012bc3a6a7dcdc

aduth commented on client/account-recovery/controller.js:
```diff
@@ -0,0 +1,32 @@
+
+		analytics.pageView.record( basePath, `${ ANALYTICS_PAGE_TITLE } > Lost Password` );
+
+		renderWithReduxStore(
```
To my previous point about using the isomorphic router, an advantage is not having to explicitly render into the DOM element:

https://github.com/Automattic/wp-calypso/blob/727fc973a841f28ff08dea300aa19813b518eec5/client/my-sites/types/controller.jsx#L18-L24

gwwar replied:
Oh nice! 😄

aduth commented on client/account-recovery/style.scss:
```diff
@@ -0,0 +1,6 @@
+.account-recovery__container {
+	margin: auto;
```
Do you need top and bottom to have `auto` margin? Else `margin: 0 auto;` might be more in line with what you're trying to achieve (horizontal centering).

omarjackman replied:
Done in https://github.com/Automattic/wp-calypso/pull/8955/commits/f711cfe9dc59c8dee757445809569bfbdee8b8d4

omarjackman commented on client/account-recovery/index.js:
```diff
@@ -0,0 +1,14 @@
+
+export default function() {
+	// Main route for account recovery is the lost password page
+	page( '/account-recovery', controller.lostPassword );
```
Done in 860e2d1

omarjackman commented:
Thanks for the great review @gwwar and @aduth 

I addressed all feedback and had a [question](https://github.com/Automattic/wp-calypso/pull/8955#discussion_r86065056) about server side rendering.

aduth commented on client/account-recovery/lost-password/index.jsx:
```diff
@@ -0,0 +1,115 @@
+
+export const LostPassword = localize( LostPasswordComponent );
+
+export const LostPasswordPage = localize( ( { translate, basePath } ) => (
```
Except for higher-order component wrappers, I think separate components like these should be in their own files. How about `main.jsx` or `page.jsx`? Or keep this here and move the form to a subdirectory like `lost-password-form` ?

gwwar replied:
Another reason we do this is so we can have easy component search. It gets difficult if we have a number of small ones in a single file.

omarjackman replied:
I restructured the form so that the components are in there own subdirectory in https://github.com/Automattic/wp-calypso/pull/8955/commits/d48e07a0e8e6e178397b1bc8a3fabad4b3a1417e

aduth commented on client/account-recovery/lost-password/style.scss:
```diff
@@ -0,0 +1,24 @@
+
+.lost-password__user-login-input {
+	//Use a more specific selector so we don't have to use !important to overide any css
+	&.form-text-input {
```
Why not merge this into the parent selector if it's the only styles being applied here?

```css
.lost-password__user-login-input.form-text-input {
```

omarjackman replied:
My initial intention was to be verbose since it all gets compiled into the same CSS but I changed it in https://github.com/Automattic/wp-calypso/pull/8955/commits/aa0fb82810a4cea03761e10cfcf87779ad6b8add

aduth commented on client/account-recovery/lost-password/index.jsx:
```diff
@@ -0,0 +1,115 @@
+
+						<FormInput
+							className="lost-password__user-login-input"
+							onChange={ this.onUserLoginChanged }
```
For this to be a controlled input, you should also assign `value={ this.state.userLogin }`

omarjackman replied:
fixed in https://github.com/Automattic/wp-calypso/pull/8955/commits/458701f0826dab77cdbb0a368d8d11db574baa49

omarjackman commented:
I did some restructuring as you suggested @aduth 

Thanks

gwwar commented on client/account-recovery/lost-password/index.jsx:
```diff
@@ -1,117 +1,21 @@
-export const LostPasswordPage = localize( ( { translate, basePath } ) => (
-	<Main>
+export default localize( ( { className, translate, basePath } ) => (
+	<Main className={ className }>
```
Would the main class ever change for this route? Perhaps pass through
```
const classes = classNames( 'account-recovery', className );
return ( 
    <Main className={ classes } >
     //...
```

omarjackman replied:
Done in https://github.com/Automattic/wp-calypso/pull/8955/commits/20ade721127559ebd98b6fb42d6b9ff1a1be5b80

gwwar commented on client/account-recovery/lost-password/lost-password-form/index.jsx:
```diff
@@ -0,0 +1,104 @@
+
+	onUserLoginChanged = ( event ) => {
+		this.setState( { userLogin: event.target.value } );
+	}
```
Missing semicolon

gwwar commented on client/lib/url/support.js:
```diff
@@ -1,6 +1,7 @@
 
 export default {
+	ACCOUNT_RECOVERY: `${ root }/account-recovery`,
 	ADDING_GOOGLE_APPS_TO_YOUR_SITE: `${root}/adding-g-suite-to-your-site`,
```
Could we add a TODO: to tidy up the missing spaces in `${ root }`?

omarjackman replied:
I tidied up in https://github.com/Automattic/wp-calypso/pull/8955/commits/3a40091503fdf4b7b778b1e513b9d808d1e9653d

gwwar commented:
👍  This looks good to me. I also tested and made sure the route wasn't accessible with the feature flag off.

omarjackman commented:
Thanks @gwwar, I'll rebase and merge this later.

dmsnell commented on client/account-recovery/lost-password/index.jsx:
```diff
@@ -0,0 +1,22 @@
+		<DocumentHead title={ translate( 'Lost Password ‹ Account Recovery' ) } />
+		<LostPasswordForm />
+	</Main>
+) );
```
Nice component!

dmsnell commented on client/account-recovery/lost-password/lost-password-form/index.jsx:
```diff
@@ -0,0 +1,104 @@
+		this.setState( { isSubmitting: true } );
+
+		//TODO: dispatch an event with userLogin and wait to here back
+	};
```
Fun fact!

You don't need the curlies here. If you leave them out then the fat-arrow becomes a function which returns the value of the expression. In this case I think it's `null` but it's whatever `this.setState()` actually returns…

```js
onSubmit = () => this.setState( { isSubmitting: true } )
```

P.S. there might be a slightly better name available which holds more meaning than `onSubmit`, such as `submitForm`. It's somewhat murky here but in general we like to avoid `onDasher` `onDancer` `onComet` and so on for handler names.

omarjackman replied:
Renamed to `submitForm` in https://github.com/Automattic/wp-calypso/pull/8955/commits/18084a9a7cd099fc4afe3941b02e5b044adbb677

I prefer to keep the curly braces since I have a comment in there and there will be some more logic in there in my next pr.

dmsnell commented on client/account-recovery/lost-password/lost-password-form/index.jsx:
```diff
@@ -0,0 +1,104 @@
+					</li>
+					<li>
+						{ translate(
+							'Follow instructions and be re-united with your {{strong}}WordPress.com{{/strong}} account',
```
This is totally subjective on my part, but were this me I'd probably split into two lines here, maybe there's a fifty-fifty chance that I would do so. It's a long line.

dmsnell commented on client/account-recovery/lost-password/lost-password-form/index.jsx:
```diff
@@ -0,0 +1,104 @@
+	}
+}
+
+export default localize( LostPasswordFormComponent );
```
@omarjackman I really appreciate how well your HTML semantically matches the content: the ordered list and h2 elements, for example. HTML really does provide some great elements and I lament how often we (myself included for sure) tend to jump to `<div>` and `<span>` for everything.

dmsnell commented on client/account-recovery/lost-password/lost-password-form/test/index.jsx:
```diff
@@ -0,0 +1,48 @@
+
+describe( 'LostPassword', () => {
+	it( 'should render as expected', () => {
+		const wrapper = shallow( <LostPasswordFormComponent translate={ identity } /> );
```
If we put `translate` in `defaultProps` for the component then we won't have to inject it everywhere, such as here in the tests.

omarjackman replied:
Fixed in https://github.com/Automattic/wp-calypso/pull/8955/commits/5400f382afc1479f0340cbd2a60f953d97d3ccc2

dmsnell commented:
@omarjackman this is a great PR, probably a good example for others to read over: well-designed stateless-first components, great tests, etc…

cc: @mtias @aduth I think this is a good example PR for others to learn from, what do you think?

mtias commented on client/account-recovery/index.js:
```diff
@@ -0,0 +1,9 @@
+/**
```
What do you think about moving this folder under `/client/auth`?

omarjackman replied:
@mtias do you think I'd need to use the [auth controller](https://github.com/Automattic/wp-calypso/blob/master/client/auth/controller.js) or it'd be fine to have them separate and keep the current account recovery controller?

mtias commented on client/account-recovery/lost-password/lost-password-form/index.jsx:
```diff
@@ -0,0 +1,104 @@
+				<h2 className="lost-password-form__title">
+					{ translate( 'Lost your password' ) }
+				</h2>
+
```
We tend to remove these empty line breaks.

omarjackman replied:
Fixed in https://github.com/Automattic/wp-calypso/pull/8955/commits/c9ecb42ffd26cb51a0f6e68fd1f32fde092ff7e9

mtias commented on client/account-recovery/lost-password/lost-password-form/style.scss:
```diff
@@ -0,0 +1,22 @@
+.lost-password-form__title {
+	font-size: 1.5em;
```
We prefer to set font-size values in `px` for consistency in general.

omarjackman replied:
Fixed in https://github.com/Automattic/wp-calypso/pull/8955/commits/0067de58d62d5c49aa7e125b67b67e83d8863aa8

mtias commented on client/account-recovery/lost-password/lost-password-form/style.scss:
```diff
@@ -0,0 +1,22 @@
+.lost-password-form__title {
+	font-size: 1.5em;
+	font-weight: bold;
```
For our system font usages (sans-serif) we need to specify `weight: 600` since we use semibold and not bold.

omarjackman replied:
Fixed in https://github.com/Automattic/wp-calypso/pull/8955/commits/8c33b88f62f87d79ed100ebfe60566134ff232d0

aduth replied:
@mtias Do you think it'd be a good idea to add `$bold: 600` as a variable to `_typography.scss` to simplify this and have it be a single point of updating if this ever changes? Usage being `font-weight: $bold;` which I think reads nicely too.

mtias commented on client/account-recovery/lost-password/lost-password-form/style.scss:
```diff
@@ -0,0 +1,22 @@
+}
+
+.lost-password-form__instruction-list {
+	margin: 10px 0px 10px 20px;
```
Minor, but we do all our margins in increments of 8 generally. Same with the 5px margin above, could be 4px instead. cc @evilluendas

omarjackman commented on client/account-recovery/controller.js:
```diff
@@ -0,0 +1,27 @@
+import { getCurrentUser } from 'state/current-user/selectors';
+
+export function lostPassword( context, next ) {
+	context.primary = <LostPasswordPage basePath={ context.path } className="account-recovery__container" />;
```
> The LostPasswordPage component is rendering Main with account-recovery class, we shouldn't have both account-recovery account-recovery__container on the same element.

Its actually rendered with `lost-password account-recovery__container main`

gwwar replied:
It feels like `account-recovery__container` is unnecessary here and also confusing since we don't have an account recovery component.

@mtias which do we prefer more?
```
.lost-password.main {
    max-width: 480px;
    font-size: 13px;
}
```
or 
```
.is-section-account-recovery .main {
    max-width: 480px;
    font-size: 13px;
}
```

omarjackman replied:
poke @mtias

omarjackman replied:
I switched it `is-section-account-recovery .main` in https://github.com/Automattic/wp-calypso/pull/8955/commits/834730a3a40cba5297d1ab8af00aba81dc54af3a

mtias commented:
@dmsnell it may be worth having a document or a place to reference good examples. The only concern is how you can be sure something doesn't become outdated (process vs implementation).

dmsnell commented:
> how you can be sure something doesn't become outdated

a link to the code and/or PR can guard that.

fun fact! if you are viewing code in Github and want to link to it, press <kbd>y</kbd> and it will reload the page affixing the currently-viewed commit/version of the file to the URL.

omarjackman commented:
@gwwar @dmsnell or @mtias can I get a ship?

gwwar commented:
:ship: :ship: :ship:
